lib/galaxy/webapps/galaxy/api/workflows.py

```diff
@@ -808,12 +810,24 @@ def invoke(self, trans: GalaxyWebTransaction, workflow_id, payload, **kwd):
         if not is_batch and len(run_configs) != 1:
             raise exceptions.RequestParameterInvalidException("Must specify 'batch' to use batch parameters.")
 
-        tool_ids = self.workflow_contents_manager.get_all_tool_ids(workflow)
-        missing_tool_ids = [tool_id for tool_id in tool_ids if not self.app.toolbox.has_tool(tool_id)]
-        if missing_tool_ids:
-            raise exceptions.MessageException(
-                f"Workflow was not invoked; the following required tools are not installed: {', '.join(missing_tool_ids)}"
+        require_exact_tool_versions = util.string_as_bool(payload.get("require_exact_tool_versions", "true"))
+        tools = self.workflow_contents_manager.get_all_tools(workflow)
+        missing_tools = [
+            tool
+            for tool in tools
+            if not self.app.toolbox.has_tool(
+                tool["tool_id"], tool_version=tool["tool_version"], exact=require_exact_tool_versions
```

Comment:
Although for local tools we're playing with fire, since we don't consistently ship copies of older versions. We only do that when the tool parameters have changed in an incompatible way.